### PR TITLE
[Flow] Allow element-wise fusion of multi-reduction ops

### PR DESCRIPTION
**Pull request by Max191**

### CODE REVIEW

**Comment by Max191** [comment]
Some element-wise fusions seen in SDXL (discussed here https://github.com/openxla/iree/issues/16450) are not happening because of a restriction on multi-reduction ops. There were some issues with doing this before (https://github.com/openxla/iree/issues/13285), but we will try this now and see what happens.

**Comment by Max191** [comment]
Will add tests (and fix failures) tomorrow for this.

**Review by qedawkins — APPROVED** [review]
This may need to be guarded by a flag; can you add a quick flag (fine to have default to true right now) that we can toggle after the fact? I remember some downstream cases of this failing in codegen.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp [hunk 1]
```diff
@@ -115,11 +116,13 @@ static bool areFusableOps(MLIRContext *context, OpOperand *fusedOperand) {
...
     }
-    if (linalgConsumerOp.getNumReductionLoops() != 1 ||
-        !linalgConsumerOp.getMatchingIndexingMap(fusedOperand)
+    if (!linalgConsumerOp.getMatchingIndexingMap(fusedOperand)
              .isPermutation()) {
       return false;
     }
+    if (linalg::isaContractionOpInterface(linalgConsumerOp)) {
+      return false;
+    }
```
Comment (qedawkins) [inline compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:130]: also block on `isaConvolutionOpInterface` here

**Review by qedawkins — CHANGES_REQUESTED** [review]

**Review by qedawkins — APPROVED** [review]
LGTM! I would wait for @MaheshRavishankar to take a look now though.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp (hunk 1 above)
Comment (hanhanW) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:130]: One of these should be `isaConvolutionOpInterface`?

**Review by hanhanW — COMMENTED** [review]
Why do we have regressions on number of dispatch launches for EfficientNetV2STF if we are fusing more?

**Review by MaheshRavishankar — APPROVED** [review]
Looks good! Thanks for plumbing this through. The flag might not be needed, but we can have it for now.

File: compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp (hunk 1 above)
Comment (Max191) [reply compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp:130]: Oh, what. Thanks for catching that

**Comment by MaheshRavishankar** [comment]
> Why do we have regressions on number of dispatch launches for EfficientNetV2STF if we are fusing more?

Yeah, but increases by 1. Worth looking at, but not a blocker for landing IMO

**Comment by hanhanW** [comment]
yeah.. not a blocker because it generally looks good. It would be good to figure it out.

**Pull request merged**